File: erlang/seminars/avl/src/avl.erl
-module(avl).

-export([tree/0, insert/3, lookup/2,  traverse/1, depth/2, max_depth/1]).

tree() ->
    nil.

insert(Tree, Key, Value) ->
    case insrt(Tree, Key, Value) of
	{inc, Q} ->
	    Q;
	{ok, Q} ->
	    Q
    end.



lookup(nil, _) ->
    fail;
lookup({node, K, V, _, _, _}, K) ->
    {ok, V};
lookup({node, K, _, _, L, R}, Key)  ->
    if
	Key < K -> lookup(L, Key);
	true -> lookup(R, Key)
    end.


depth(nil, _) ->
    fail;
depth({node, K, _, _, _, _}, K) ->
    {ok, 1};
depth({node, K, _, _, L, R}, Key)  ->
    if
	Key < K -> 
	    case depth(L, Key) of
		{ok, D} -> {ok, D + 1};
		fail -> fail
	    end;
	true -> 
	    case depth(R, Key) of
		{ok, D} -> {ok, D + 1};
		fail -> fail
	    end
    end.


max_depth(nil) ->
    0;
max_depth({node, _, _, _, L, R})  ->
    max(max_depth(L), max_depth(R)) + 1.


traverse(nil) ->
    {0,0};
traverse({node, _, _, _, L, R}) ->
    {Lt, Lm} = traverse(L),
    {Rt, Rm} = traverse(R),    
    {Lt+Rt+1, max(Lm, Rm)+1}.



insrt(nil, Key, Value) ->
    %% empty tree
    {inc, {node, Key, Value, 0, nil, nil}};

insrt({node, Key, _, F, A, B}, Key, Value) ->
    %% found in root
    {ok, {node, Key, Value, F, A, B}};


insrt({node, Rk, Rv, 0, A, B}, Kk, Kv)  when Kk < Rk ->
    case insrt(A, Kk, Kv) of
	{inc, Q} ->
	    {inc, {node, Rk, Rv, -1, Q, B}};
	{ok, Q} ->
	    {ok, {node, Rk, Rv, 0, Q, B}}
    end;

insrt({node, Rk, Rv, 0, A, B}, Kk, Kv)  when Kk > Rk ->
    case insrt(B, Kk, Kv) of
	{inc, Q} ->
	    {inc, {node, Rk, Rv, +1, A, Q}};
	{ok, Q} ->
	    {ok, {node, Rk, Rv, 0, A, Q}}
    end;

insrt({node, Rk, Rv, +1, A, B}, Kk, Kv)  when Kk < Rk ->
    case insrt(A, Kk, Kv) of
	{inc, Q} ->
	    {ok, {node, Rk, Rv, 0, Q, B}};
	{ok, Q} ->
	    {ok, {node, Rk, Rv, +1, Q, B}}
    end;

insrt({node, Rk, Rv, -1, A, B}, Kk, Kv)  when Kk > Rk ->
    case insrt(B, Kk, Kv) of
	{inc, Q} ->
	    {ok, {node, Rk, Rv, 0, A, Q}};
	{ok, Q} ->
	    {ok, {node, Rk, Rv, -1, A, Q}}
    end;

insrt({node, Rk, Rv, -1, A, B}, Kk, Kv)  when Kk < Rk ->
    case insrt(A, Kk, Kv) of
	{inc, Q} ->
	    {ok, rotate({node, Rk, Rv, -2, Q, B})};
	{ok, Q} ->
	    {ok, {node, Rk, Rv, -1, Q, B}}
    end;

insrt({node, Rk, Rv, +1, A, B}, Kk, Kv)  when Kk > Rk ->
    case insrt(B, Kk, Kv) of
	{inc, Q} ->
	    {ok, rotate({node, Rk, Rv, +2, A, Q})};
	{ok, Q} ->
	    {ok, {node, Rk, Rv, +1, A, Q}}
    end.


%% left - singel rotate

rotate({node, Xk, Xv, -2, {node, Yk, Yv, -1, A, B}, C}) ->	    
    %% rotate 
    {node, Yk, Yv, 0, A, {node, Xk, Xv, 0, B, C}};

%% right - singel rotate

rotate({node, Xk, Xv, +2, A, {node, Yk, Yv, +1, B, C}}) ->	    
    %% rotate 
    {node, Yk, Yv, 0, {node, Xk, Xv, 0, A, B}, C};


%% left - double rotate 

rotate({node, Xk, Xv, -2, {node, Yk, Yv, +1, A, {node, Zk, Zv, 0, B, C}}, D}) ->	    
    %% double rotate 
    {node, Zk, Zv, 0, {node, Yk, Yv, 0, A, B}, {node, Xk, Xv, 0, C, D}};

rotate({node, Xk, Xv, -2, {node, Yk, Yv, +1, A, {node, Zk, Zv, -1, B, C}}, D}) ->	    
    %% double rotate 
    {node, Zk, Zv, 0, {node, Yk, Yv, 0, A, B}, {node, Xk, Xv, +1, C, D}};

rotate({node, Xk, Xv, -2, {node, Yk, Yv, +1, A, {node, Zk, Zv, +1, B, C}}, D}) ->	    
    %% double rotate 
    {node, Zk, Zv, 0, {node, Yk, Yv, -1, A, B}, {node, Xk, Xv, 0, C, D}};


%% right - double rotate

rotate({node, Xk, Xv, +2, A, {node, Yk, Yv, -1, {node, Zk, Zv, 0, B, C}, D}}) ->	    
    %% double rotate 
    {node, Zk, Zv, 0, {node, Xk, Xv, 0, A, B}, {node, Yk, Yv, 0, C, D}};

rotate({node, Xk, Xv, +2, A, {node, Yk, Yv, -1, {node, Zk, Zv, +1, B, C}, D}}) ->	    
    %% double rotate 
    {node, Zk, Zv, 0, {node, Xk, Xv, -1, A, B}, {node, Yk, Yv, 0, C, D}};

rotate({node, Xk, Xv, +2, A, {node, Yk, Yv, -1, {node, Zk, Zv, -1, B, C}, D}}) ->	    
    %% double rotate 
    {node, Zk, Zv, 0, {node, Xk, Xv, 0, A, B}, {node, Yk, Yv, +1, C, D}}.

File: exercises/avl/src/avl.ex
defmodule AVL do

  # ================================================== #
  # == AVL Tree - Public API == #
  # ================================================== #

  # Create an empty tree.
  def tree(), do: nil

  # Insert a key-value pair in a tree.
  # Usage: insert(tree, key, value)
  def insert(tree, key, value) do
    case insrt(tree, key, value) do
      {:inc, q} -> q
      {:ok, q} -> q
    end
  end

  # Lookup for a given key in a tree.
  # Usage: lookup(tree, key)
  def lookup(nil, _), do: :fail
  def lookup({:node, k, v, _, _, _}, k), do: {:ok, v}
  def lookup({:node, k, _, _, l, r}, key) do
    cond do
      key < k -> lookup(l, key)
      true -> lookup(r, key)
    end
  end

  # Compute the depth of a the leaf containing a given key
  # in a tree.
  # Usage: depth(tree, key)
  def depth(nil, _), do: :fail
  def depth({:node, k, _, _, _, _}, k), do: {:ok, 1}
  def depth({:node, k, _, _, l, r}, key) do
    cond do
      key < k ->
        case depth(l, key) do
          {:ok, d} -> {:ok, d + 1}
          :fail -> :fail
        end

      true ->
        case depth(r, key) do
          {:ok, d} -> {:ok, d + 1}
          :fail -> :fail
        end
    end
  end

  # Compute the maximum depth of a tree.
  # Usage: max_depth(tree)
  def max_depth(nil), do: 0
  def max_depth({:node, _, _, _, l, r}) do
    max(max_depth(l), max_depth(r)) + 1
  end

  # Traverse the tree to compute a tuple containing the 
  # number of nodes and the maximum depth.
  # Usage: traverse(tree)
  def traverse(nil), do: {0, 0}
  def traverse({:node, _, _, _, l, r}) do
    {lt, lm} = traverse(l)
    {rt, rm} = traverse(r)
    {lt + rt + 1, max(lm, rm) + 1}
  end

  # ================================================== #
  # == Private methods == #
  # ================================================== #

  defp insrt(nil, key, value) do
    {:inc, {:node, key, value, 0, nil, nil}}
  end
  defp insrt({:node, key, _, f, a, b}, key, value) do
    {:ok, {:node, key, value, f, a, b}}
  end
  defp insrt({:node, rk, rv, 0, a, b}, kk, kv) when kk < rk do
    case insrt(a, kk, kv) do
      {:inc, q} ->
        {:inc, {:node, rk, rv, -1, q, b}}

      {:ok, q} ->
        {:ok, {:node, rk, rv, 0, q, b}}
    end
  end
  defp insrt({:node, rk, rv, 0, a, b}, kk, kv) when kk > rk do
    case insrt(b, kk, kv) do
      {:inc, q} ->
        {:inc, {:node, rk, rv, +1, a, q}}

      {:ok, q} ->
        {:ok, {:node, rk, rv, 0, a, q}}
    end
  end
  defp insrt({:node, rk, rv, +1, a, b}, kk, kv) when kk < rk do
    case insrt(a, kk, kv) do
      {:inc, q} ->
        {:ok, {:node, rk, rv, 0, q, b}}

      {:ok, q} ->
        {:ok, {:node, rk, rv, +1, q, b}}
    end
  end
  defp insrt({:node, rk, rv, -1, a, b}, kk, kv) when kk > rk do
    case insrt(b, kk, kv) do
      {:inc, q} ->
        {:ok, {:node, rk, rv, 0, a, q}}

      {:ok, q} ->
        {:ok, {:node, rk, rv, -1, a, q}}
    end
  end
  defp insrt({:node, rk, rv, -1, a, b}, kk, kv) when kk < rk do
    case insrt(a, kk, kv) do
      {:inc, q} ->
        {:ok, rotate({:node, rk, rv, -2, q, b})}

      {:ok, q} ->
        {:ok, {:node, rk, rv, -1, q, b}}
    end
  end
  defp insrt({:node, rk, rv, +1, a, b}, kk, kv) when kk > rk do
    case insrt(b, kk, kv) do
      {:inc, q} ->
        {:ok, rotate({:node, rk, rv, +2, a, q})}

      {:ok, q} ->
        {:ok, {:node, rk, rv, +1, a, q}}
    end
  end

  # Single rotate left
  defp rotate({:node, xk, xv, -2, {:node, yk, yv, -1, a, b}, c}) do
    {:node, yk, yv, 0, a, {:node, xk, xv, 0, b, c}}
  end

  # Single rotate right
  defp rotate({:node, xk, xv, +2, a, {:node, yk, yv, +1, b, c}}) do
    {:node, yk, yv, 0, {:node, xk, xv, 0, a, b}, c}
  end

  # Double rotate left
  defp rotate({:node, xk, xv, -2, {:node, yk, yv, +1, a, {:node, zk, zv, 0, b, c}}, d}) do
    {:node, zk, zv, 0, {:node, yk, yv, 0, a, b}, {:node, xk, xv, 0, c, d}}
  end
  defp rotate({:node, xk, xv, -2, {:node, yk, yv, +1, a, {:node, zk, zv, -1, b, c}}, d}) do
    {:node, zk, zv, 0, {:node, yk, yv, 0, a, b}, {:node, xk, xv, +1, c, d}}
  end
  defp rotate({:node, xk, xv, -2, {:node, yk, yv, +1, a, {:node, zk, zv, +1, b, c}}, d}) do
    {:node, zk, zv, 0, {:node, yk, yv, -1, a, b}, {:node, xk, xv, 0, c, d}}
  end

  # Double rotate right
  defp rotate({:node, xk, xv, +2, a, {:node, yk, yv, -1, {:node, zk, zv, 0, b, c}, d}}) do
    {:node, zk, zv, 0, {:node, xk, xv, 0, a, b}, {:node, yk, yv, 0, c, d}}
  end
  defp rotate({:node, xk, xv, +2, a, {:node, yk, yv, -1, {:node, zk, zv, +1, b, c}, d}}) do
    {:node, zk, zv, 0, {:node, xk, xv, -1, a, b}, {:node, yk, yv, 0, c, d}}
  end
  defp rotate({:node, xk, xv, +2, a, {:node, yk, yv, -1, {:node, zk, zv, -1, b, c}, d}}) do
    {:node, zk, zv, 0, {:node, xk, xv, 0, a, b}, {:node, yk, yv, +1, c, d}}
  end
end
